athena/etc/larvnetd/larvnetd.8
.\" $Id: larvnetd.8,v 1.1 1998-09-01 20:57:44 ghudson Exp $
.\"
.\" Copyright 1998 by the Massachusetts Institute of Technology.
.\"
.\" Permission to use, copy, modify, and distribute this
.\" software and its documentation for any purpose and without
.\" fee is hereby granted, provided that the above copyright
.\" notice appear in all copies and that both that copyright
.\" notice and this permission notice appear in supporting
.\" documentation, and that the name of M.I.T. not be used in
.\" advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.
.\" M.I.T. makes no representations about the suitability of
.\" this software for any purpose.  It is provided "as is"
.\" without express or implied warranty.
.TH LARVNETD 8 "23 Aug 1998"
.SH NAME
larvnetd \- Monitor cluster workstations and printers
.SH SYNOPSIS
.B larvnetd [-n] [-f \fIconfigfile\fR]
.SH DESCRIPTION
.B larvnetd
monitors a set of cluster machines and printers given by the file
.B /etc/athena/larvnet.conf
(see
.BR larvnet.conf (5))
or a file specified on the command line with the \-f option.
Normally, it backgrounds itself and dissociates itself from the
controlling tty after binding to the larvnet port, but this behavior
can be suppressed with the \-n command-line option.
.PP
.B larvnetd
processes status notifications received from cluster workstations on
the larvnet port.  It will periodically poll those workstations it has
not heard from recently, using the busypoll service (see
.BR busyd (8)).
Once a minute,
.B larvnetd
requests a queue status for each of the printers it is configured to
monitor.  Also once a minute,
.B larvnetd
generates three report files:
.RS
.B /var/athena/larvnet.cgroups
.br
.B /var/athena/larvnet.clusters
.br
.B /var/athena/larvnet.printers
.RE
which give the current known status of cluster workstations and
printers.
.PP
.B larvnetd
will reread its configuration file upon receiving a SIGHUP.
.PP
.B larvnetd
syslogs errors and debugging messages using the local5 facility name.
.SH NOTES
Larvnet is monitoring you.
.SH FILES
.B /etc/athena/larvnet.conf
.br
.B /var/athena/larvnet.cgroups
.br
.B /var/athena/larvnet.clusters
.br
.B /var/athena/larvnet.printers
.SH "SEE ALSO"
.BR larvnet.conf (5),
.BR larvnet.cgroups (5),
.BR larvnet.clusters (5),
.BR larvnet.printers (5),
.BR busyd (8)
.SH AUTHOR
Greg Hudson, MIT Information Systems
.br
Copyright 1998 by the Massachusetts Institute of Technology.


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


athena/etc/larvnetd/larvnetd.c
/* Copyright 1998 by the Massachusetts Institute of Technology.
 *
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright
 * notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in
 * advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 */

/* This file is part of larvnetd, a monitoring server.  It implements
 * the main loop.
 */

static const char rcsid[] = "$Id: larvnetd.c,v 1.4 1999-10-19 20:23:31 danw Exp $";

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <netdb.h>
#include <syslog.h>
#include <hesiod.h>
#include "larvnetd.h"
#include "larvnet.h"
#include "timer.h"

extern char *optarg;
extern int optind;

static volatile int reload;

static void usage(const char *progname);
static void hup_handler(void);

int main(int argc, char **argv)
{
  struct timeval tv, *tvp;
  struct sockaddr_in sin;
  struct sigaction action;
  struct servent *servent;
  struct serverstate state;
  struct printer *printer;
  unsigned short port;
  pid_t pid;
  fd_set readers, writers;
  int c, nofork = 0, fd, i, nfds, count, status;
  const char *progname;
  char *errmem;
  FILE *fp;

  progname = strrchr(argv[0], '/');
  progname = (progname) ? progname + 1 : argv[0];

  state.configfile = LARVNET_PATH_CONFIG;
  while ((c = getopt(argc, argv, "nf:")) != -1)
    {
      switch (c)
	{
	case 'n':
	  nofork = 1;
	  break;
	case 'f':
	  state.configfile = optarg;
	  break;
	default:
	  usage(progname);
	}
    }
  if (argc != optind)
    usage(progname);

  /* Bind to the server socket. */
  servent = getservbyname("larvnet", "udp");
  port = (servent) ? servent->s_port : htons(LARVNET_FALLBACK_PORT);
  state.server_socket = socket(AF_INET, SOCK_DGRAM, 0);
  if (state.server_socket == -1)
    {
      fprintf(stderr, "%s: socket: %s\n", progname, strerror(errno));
      return 1;
    }
  memset(&sin, 0, sizeof(sin));
  sin.sin_family = AF_INET;
  sin.sin_port = port;
  if (bind(state.server_socket, (struct sockaddr *) &sin, sizeof(sin)) == -1)
    {
      fprintf(stderr, "%s: bind: %s\n", progname, strerror(errno));
      return 1;
    }

  /* Initialize a resolver channel. */
  status = ares_init(&state.channel);
  if (status != ARES_SUCCESS)
    {
      fprintf(stderr, "Can't initialize resolver: %s\n",
	      ares_strerror(status, &errmem));
      ares_free_errmem(errmem);
      return 1;
    }

  /* Initialize a Hesiod context. */
  if (hesiod_init(&state.hescontext) == -1)
    {
      fprintf(stderr, "Can't initialize Hesiod context: %s\n",
	      strerror(errno));
      return 1;
    }

  /* Determine the port for busy status polls. */
  servent = getservbyname("busypoll", "udp");
  state.poll_port = (servent) ? servent->s_port
      : htons(BUSYPOLL_FALLBACK_PORT);

  if (!nofork)
    {
      /* fork, dissociate from our tty, chdir to /, repoint
       * stdin/stdout/stderr at /dev/null.  Standard daemon stuff.
       */
      pid = fork();
      if (pid == -1)
	{
	  fprintf(stderr, "%s: fork: %s", progname, strerror(errno));
	  return 1;
	}
      if (pid != 0)
	_exit(0);
      if (setsid() == -1)
	{
	  fprintf(stderr, "%s: setsid: %s", progname, strerror(errno));
	  return 1;
	}
      chdir("/");
      fd = open("/dev/null", O_RDWR, 0);
      if (fd != -1)
	{
	  dup2(fd, STDIN_FILENO);
	  dup2(fd, STDOUT_FILENO);
	  dup2(fd, STDERR_FILENO);
	  if (fd != STDIN_FILENO && fd != STDOUT_FILENO && fd != STDERR_FILENO)
	    close(fd);
	}

      /* Write a pid file. */
      fp = fopen(LARVNET_PATH_PIDFILE, "w");
      if (fp)
	{
	  fprintf(fp, "%lu\n", (unsigned long) getpid());
	  fclose(fp);
	}
    }

  /* Set signal handlers: reread the configuration file on HUP, and
   * ignore SIGPIPE because every robust network program's gotta.
   */
  sigemptyset(&action.sa_mask);
  action.sa_flags = 0;
  action.sa_handler = hup_handler;
  sigaction(SIGHUP, &action, NULL);
  action.sa_handler = SIG_IGN;
  sigaction(SIGPIPE, &action, NULL);

  /* Start syslogging. */
  openlog(progname, LOG_PID, LOG_LOCAL5);

  read_initial_config(&state);

  /* Start initial polls of workstations and printers, and start the
   * report generation timer. */
  state.startmachine = 0;
  ws_poll(&state);
  printer_start_polls(&state);
  timer_set_rel(60, report, &state.config);

  while (1)
    {
      /* See if we got a SIGHUP and need to reread the config file. */
      if (reload)
	{
	  reload = 0;
	  reread_config(&state);
	  printer_start_polls(&state);
	}

      /* Set up fd sets for select. */
      FD_ZERO(&readers);
      FD_ZERO(&writers);
      nfds = ares_fds(state.channel, &readers, &writers);
      FD_SET(state.server_socket, &readers);
      if (state.server_socket >= nfds)
	nfds = state.server_socket + 1;
      for (i = 0; i < state.config.nprinters; i++)
	{
	  printer = &state.config.printers[i];
	  if (printer->s == -1)
	    continue;
	  if (printer->to_send)
	    FD_SET(printer->s, &writers);
	  else
	    FD_SET(printer->s, &readers);
	  if (printer->s >= nfds)
	    nfds = printer->s + 1;
	}

      /* Find the minimum of the ares and timer timeouts. */
      tvp = timer_timeout(&tv);
      tvp = ares_timeout(state.channel, tvp, &tv);

      if (tvp)
	{
	  syslog(LOG_DEBUG, "main: select timeout %lu %lu",
		 (unsigned long) tvp->tv_sec, (unsigned long) tvp->tv_usec);
	}
      else
	syslog(LOG_DEBUG, "main: no select timeout");

      /* Wait for network input or a timer event. */
      count = select(nfds, &readers, &writers, NULL, tvp);
      if (count < 0 && errno != EINTR)
	{
	  syslog(LOG_ALERT, "main: aborting on select error: %m");
	  return 1;
	}

      syslog(LOG_DEBUG, "main: select count %d", count);

      ares_process(state.channel, &readers, &writers);
      timer_process();

      /* Don't bother scanning fd sets if nothing happened. */
      if (count <= 0)
	continue;

      if (FD_ISSET(state.server_socket, &readers))
	ws_handle_status(state.server_socket, &state.config);
      for (i = 0; i < state.config.nprinters; i++)
	{
	  printer = &state.config.printers[i];
	  if (printer->s == -1)
	    continue;
	  if (FD_ISSET(printer->s, &readers))
	    printer_handle_input(&state, printer);
	  else if (FD_ISSET(printer->s, &writers))
	    printer_handle_output(&state, printer);
	}
    }
}

static void usage(const char *progname)
{
  fprintf(stderr, "Usage: %s [-n] [-f configfile]\n", progname);
  exit(1);
}

static void hup_handler(void)
{
  reload = 1;
}
